### ISSUE TITLE
Invalid combination of opcode and operands gets compiled

### ISSUE THREAD

**Issue by Leaeaxedx** [Terraspace/UASM#113]
This
  VMOVHPD   xmm2,xmm0,xmm1

or even

 VMOVHPD   xmm2,xmm0,rcx

gets compiled. The third operands can only be a mem64.

Same goes for VMOVLPD and VMOVLPS/VMOVHPS.

**Comment by john-terraspace** [comment]
Hi,

 

These issues (along with many other code-gen issues like this in terms of proper validation) are all being handled by migrating to the new V2 CodeGen.

I will keep this item here so we can ensure there are new regression/test-cases for the V2 gen. The new instruction support is not yet in Git, but we have about 150 instructions translated so far.

 

John

**Comment by Leaeaxedx** [comment]
V2 CodeGen?

BTW, thanks for UASM. Nice job.

V V sob., 17. avg. 2019 ob 21:41 je oseba John Hankinson <
notifications@github.com> napisala:

**Comment by john-terraspace** [comment]
Yep, If you look in the source the original codegen.c (which has been in use since wasm) has been sticky taped and tweaked over years. I eventually gave up on it in January, it’s too fragile to keep adding manual fixes instruction by instruction. So now there is codegenv2.c. The parser tries V2 first, only if it can’t find the instruction does it fall back to the original one, just to ensure we have complete instruction support while migrating.

**Comment by Leaeaxedx** [comment]
If I understand it correctly some portion of code is in re-writing process? Any time line?

Regards, Igor

**Comment by john-terraspace** [comment]
Hi,

 

The Code Gen is being re-written, 2.49 and 2.50 already make extensive use of the new code-gen however it’s been designed to fall back to the old one for instructions not yet implemented so we can transition slowly/carefully. 

 

That’s about the only part at present which is a complete re-build.

 

I will probably be replacing the ELF 64bit output stage too to properly support PLT/GOT and Dwarf.

 

John

**Comment by Leaeaxedx** [comment]
I see. Thanks.

Have a nice coding.

V V tor., 24. sep. 2019 ob 20:20 je oseba John Hankinson <
notifications@github.com> napisala:

**Comment by Leaeaxedx** [comment]
Is this fixed now?

If so where to get the latest UASM exe?

Regards,

Igor

V V pet., 26. feb. 2021 ob 17:27 je oseba John Hankinson <
notifications@github.com> napisala:

**Comment by john-terraspace** [comment]
Accidentally closed this, I've just released 2.51 (git and on the website). I've moved some more instructions into CodeGen V2.
The WIP folder has a version with "most" instructions already moved, but I do them in small chunks as each one needs it's own regression tests and is it goes across.

**Comment by Leaeaxedx** [comment]
Hi John,

hate to bring bad news, but I found out

(v)pmovzx?/? doesn't compile when the destination is a zmm register

like  vpmovzxbd zmm0, xmm0 or xmmword ptr [address]

Same goes for  (v)pmovsx?/? This is true for 2.50 as 2.51 version.

Have a nice day.

Igor

V V sob., 27. feb. 2021 ob 23:01 je oseba John Hankinson <
***@***.***> napisala:

**Comment by Leaeaxedx** [comment]
Sorry, John, my bad.

I was getting the compile errors
 1. DATA segment with different attributes (with only one DATA segment in
the project!)
  2. VCEnd...with some negative number at the end
 3. (v)pmovzx and (v)pmovsx invalid operands

in the above order. I couldn't fix it no matter what.

Today all errors went away. I can not reproduce them. I don't know why this
happened.

It looks like restarting the windows helped.

Sorry to bother you. Keep up the good work.

Regards,

Igor

V V pon., 22. mar. 2021 ob 12:42 je oseba Igor Fujs ***@***.***>
napisala:
